### PR TITLE
Cythonize concat method

### PR DESCRIPTION
**Pull request by unnonouno**
Please merge #2144 before this pr.
Concat method is slow because it consumes to check arguments. I cythonized concat method.

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1983,19 +1983,145 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
...
+            shape = a._shape
+            if axis < 0:
+                axis += ndim
+            if axis < 0 or axis >= ndim:
+                raise IndexError(
+                    'axis {} out of bounds [0, {})'.format(axis, ndim))
+            dtype = a.dtype
+            continue
+
+        have_same_types &= (a.dtype == dtype)
```
Comment (okuta) [inline cupy/core/core.pyx]: Please use `and`

**Comment by unnonouno** [comment]
Current array comparison is faster than `shape[:axis]` and `shape[axis+1:]` because cython optimizes it.

**Comment by unnonouno** [comment]
I now get this result. The original implementation requires about 15ms. It is 10times faster!
```
CPU split_axis ... finished in 5.18ms
CPU concat ... finished in 3.27ms
copy to GPU ... finished in 6.92ms
GPU split_axis ... finished in 4.88ms
GPU concat ... finished in 1.15ms
copy to CPU ... finished in 1.21ms
```

**Comment by okuta** [comment]
LGTM!

**Pull request merged**